File: dotnet_projects/OcrLiteOnnxVB/OcrLiteLib/CrnnNet.vb
Imports System.Drawing
Imports System.IO
Imports System.Text
Imports Emgu.CV
Imports Microsoft.ML.OnnxRuntime
Imports Microsoft.ML.OnnxRuntime.Tensors

Namespace OcrLiteLib
    Public Class CrnnNet
        Private ReadOnly MeanValues As Single() = {127.5F, 127.5F, 127.5F}
        Private ReadOnly NormValues As Single() = {1.0F / 127.5F, 1.0F / 127.5F, 1.0F / 127.5F}
        Private Const crnnDstHeight As Integer = 32
        Private Const crnnCols As Integer = 5531

        Private crnnNet As InferenceSession
        Private keys As List(Of String)

        Public Sub New(path As String, keysPath As String, numThread As Integer)
            Try
                Dim op As New SessionOptions()
                op.GraphOptimizationLevel = GraphOptimizationLevel.ORT_ENABLE_EXTENDED
                op.InterOpNumThreads = numThread
                op.IntraOpNumThreads = numThread
                crnnNet = New InferenceSession(path, op)
                keys = InitKeys(keysPath)
            Catch ex As Exception
                Console.WriteLine(ex.Message + ex.StackTrace)
                Throw ex
            End Try
        End Sub

        Protected Overrides Sub Finalize()
            Try
                crnnNet.Dispose()
            Finally
                MyBase.Finalize()
            End Try
        End Sub

        Private Function InitKeys(path As String) As List(Of String)
            Dim sr As New StreamReader(path, Encoding.UTF8)
            Dim keys As New List(Of String)()
            Dim line As [String]
            While (InlineAssignHelper(line, sr.ReadLine())) IsNot Nothing
                'Console.WriteLine(line.ToString());
                keys.Add(line)
            End While
            Console.WriteLine($"keys Size = {keys.Count}")
            Return keys
        End Function

        Public Function GetTextLines(partImgs As List(Of Mat)) As List(Of TextLine)
            Dim textLines As New List(Of TextLine)
            For i As Integer = 0 To partImgs.Count - 1
                Dim startTicks = DateTime.Now.Ticks
                Dim textLine = GetTextLine(partImgs(i))
                Dim endTicks = DateTime.Now.Ticks
                Dim crnnTime = (endTicks - startTicks) / 10000.0F
                textLine.Time = crnnTime
                textLines.Add(textLine)
            Next
            Return textLines
        End Function

        Private Function GetTextLine(src As Mat) As TextLine
            Dim textLine As New TextLine()

            Dim scale As Single = CSng(crnnDstHeight) / CSng(src.Rows)
            Dim dstWidth As Integer = CInt(CSng(src.Cols) * scale)

            Dim srcResize As New Mat()
            CvInvoke.Resize(src, srcResize, New Size(dstWidth, crnnDstHeight))
            Dim inputTensors As Tensor(Of Single) = OcrUtils.SubstractMeanNormalize(srcResize, MeanValues, NormValues)
            Dim inputs = New List(Of NamedOnnxValue) From {
                NamedOnnxValue.CreateFromTensor("input", inputTensors)
            }
            Try
                Using results As IDisposableReadOnlyCollection(Of DisposableNamedOnnxValue) = crnnNet.Run(inputs)
                    Dim resultsArray = results.ToArray()
                    Console.WriteLine(resultsArray)
                    Dim outputData As Single() = resultsArray(0).AsEnumerable(Of Single)().ToArray()

                    Dim crnnRows As Integer = outputData.Count() / crnnCols

                    Return ScoreToTextLine(outputData, crnnRows, crnnCols)
                End Using
            Catch ex As Exception
                'throw ex;
                Console.WriteLine(ex.Message + ex.StackTrace)
            End Try

            Return textLine
        End Function

        Private Function ScoreToTextLine(srcData As Single(), rows As Integer, cols As Integer) As TextLine
            Dim sb As New StringBuilder()
            Dim textLine As New TextLine()

            Dim lastIndex As Integer = 0
            Dim scores As New List(Of Single)

            For i As Integer = 0 To rows - 1
                Dim maxIndex As Integer = 0
                Dim maxValue As Single = -1000.0F

                'do softmax
                Dim expList As New List(Of Single)
                For j As Integer = 0 To cols - 1
                    Dim expSingle As Single = CSng(Math.Exp(srcData(i * cols + j)))
                    expList.Add(expSingle)
                Next
                Dim partition As Single = expList.Sum()
                For j As Integer = 0 To cols - 1
                    Dim softmax As Single = expList(j) / partition
                    If softmax > maxValue Then
                        maxValue = softmax
                        maxIndex = j
                    End If
                Next

                If maxIndex > 0 AndAlso maxIndex < keys.Count AndAlso (Not (i > 0 AndAlso maxIndex = lastIndex)) Then
                    scores.Add(maxValue)
                    sb.Append(keys(maxIndex - 1))
                End If
                lastIndex = maxIndex
            Next
            textLine.Text = sb.ToString()
            textLine.CharScores = scores
            Return textLine
        End Function
        Private Shared Function InlineAssignHelper(Of T)(ByRef target As T, value As T) As T
            target = value
            Return value
        End Function
    End Class
End Namespace

File: dotnet_projects/OcrLiteOnnxCs/OcrLiteLib/CrnnNet.cs
using Emgu.CV;
using Microsoft.ML.OnnxRuntime;
using Microsoft.ML.OnnxRuntime.Tensors;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.IO;
using System.Linq;
using System.Text;

namespace OcrLiteLib
{
    class CrnnNet
    {
        private readonly float[] MeanValues = { 127.5F, 127.5F, 127.5F };
        private readonly float[] NormValues = { 1.0F / 127.5F, 1.0F / 127.5F, 1.0F / 127.5F };
        private const int crnnDstHeight = 32;
        private const int crnnCols = 5531;

        private InferenceSession crnnNet;
        private List<string> keys;

        public CrnnNet() { }

        ~CrnnNet()
        {
            crnnNet.Dispose();
        }

        public void InitModel(string path, string keysPath, int numThread)
        {
            try
            {
                SessionOptions op = new SessionOptions();
                op.GraphOptimizationLevel = GraphOptimizationLevel.ORT_ENABLE_EXTENDED;
                op.InterOpNumThreads = numThread;
                op.IntraOpNumThreads = numThread;
                crnnNet = new InferenceSession(path, op);
                keys = InitKeys(keysPath);
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message + ex.StackTrace);
                throw ex;
            }
        }
        private List<string> InitKeys(string path)
        {
            StreamReader sr = new StreamReader(path, Encoding.UTF8);
            List<string> keys = new List<string>();
            String line;
            while ((line = sr.ReadLine()) != null)
            {
                //Console.WriteLine(line.ToString());
                keys.Add(line);
            }
            Console.WriteLine($"keys Size = {keys.Count}");
            return keys;
        }

        public List<TextLine> GetTextLines(List<Mat> partImgs)
        {
            List<TextLine> textLines = new List<TextLine>();
            for (int i = 0; i < partImgs.Count; i++)
            {
                var startTicks = DateTime.Now.Ticks;
                var textLine = GetTextLine(partImgs[i]);
                var endTicks = DateTime.Now.Ticks;
                var crnnTime = (endTicks - startTicks) / 10000F;
                textLine.Time = crnnTime;
                textLines.Add(textLine);
            }
            return textLines;
        }

        private TextLine GetTextLine(Mat src)
        {
            TextLine textLine = new TextLine();

            float scale = (float)crnnDstHeight / (float)src.Rows;
            int dstWidth = (int)((float)src.Cols * scale);

            Mat srcResize = new Mat();
            CvInvoke.Resize(src, srcResize, new Size(dstWidth, crnnDstHeight));
            Tensor<float> inputTensors = OcrUtils.SubstractMeanNormalize(srcResize, MeanValues, NormValues);
            var inputs = new List<NamedOnnxValue>
            {
                NamedOnnxValue.CreateFromTensor("input", inputTensors)
            };
            try
            {
                using (IDisposableReadOnlyCollection<DisposableNamedOnnxValue> results = crnnNet.Run(inputs))
                {
                    var resultsArray = results.ToArray();
                    Console.WriteLine(resultsArray);
                    float[] outputData = resultsArray[0].AsEnumerable<float>().ToArray();

                    int crnnRows = outputData.Count() / crnnCols;

                    return ScoreToTextLine(outputData, crnnRows, crnnCols);
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message + ex.StackTrace);
                //throw ex;
            }

            return textLine;
        }

        private TextLine ScoreToTextLine(float[] srcData, int rows, int cols)
        {
            StringBuilder sb = new StringBuilder();
            TextLine textLine = new TextLine();

            int lastIndex = 0;
            List<float> scores = new List<float>();

            for (int i = 0; i < rows; i++)
            {
                int maxIndex = 0;
                float maxValue = -1000F;

                //do softmax
                List<float> expList = new List<float>();
                for (int j = 0; j < cols; j++)
                {
                    float expSingle = (float)Math.Exp(srcData[i * cols + j]);
                    expList.Add(expSingle);
                }
                float partition = expList.Sum();
                for (int j = 0; j < cols; j++)
                {
                    float softmax = expList[j] / partition;
                    if (softmax > maxValue)
                    {
                        maxValue = softmax;
                        maxIndex = j;
                    }
                }

                //no softmax
                /*for (int j = 0; j < cols; j++) {
                    if (srcData[i * cols + j] > maxValue) {
                        maxValue = srcData[i * cols + j];
                        maxIndex = j;
                    }
                }*/

                if (maxIndex > 0 && maxIndex < keys.Count && (!(i > 0 && maxIndex == lastIndex)))
                {
                    scores.Add(maxValue);
                    sb.Append(keys[maxIndex - 1]);
                }
                lastIndex = maxIndex;
            }
            textLine.Text = sb.ToString();
            textLine.CharScores = scores;
            return textLine;
        }

    }
}